### PR TITLE
[iOS] Resolve inputText flakiness

### PR DESCRIPTION
**Pull request by artem888**
## Proposed Changes

This PR implements a fix for inputText flakiness on iOS. The main reason for the flakiness is that in case of typingSpeed being too fast characters are being skipped. Can be reliably reproduced when typingSpeed is set to 30 chars per second.

New logic has several improvements:
- taps automatically on the cell, so additional tapOn command is not needed anymore. supported params (`id`, `accessibilityText`, `point`)
- in case of text length equal or less 50 characters inputText would be used as normally with reduced typingSpeed (10 characters per second)
- in case of text length exceeding 50 characters new logic of copy-paste will be used

before:
```
- tapOn:
    id: "login-email-input"
- inputText: "user@example.com"
```

after:
```
- inputText:
    text: "user@example.com"
    id: "login-email-input"
```

short text (<50 chars) input video:

https://github.com/mobile-dev-inc/maestro/assets/961175/3e778ff5-8d5d-414a-9f8b-e7de6cb51a7a

long text (>50 chars) input video:

https://github.com/mobile-dev-inc/maestro/assets/961175/78fc8dbf-55c8-424a-bb8c-e11f0cecda25

## Testing
maestro-flow-tester tests:
- [x] UIKit app
- [x] ReactNative app
- [x] Flutter app

## Issues Fixed
https://github.com/mobile-dev-inc/maestro/issues/395

### CODE REVIEW

**Comment by igorsmotto** [comment]
Hey! Great to see this PR, I have some general questions:
1. Is this API change required?
2. How it works for other platforms (i.e.: Android)?
3. Does this solution works for secureTextEntry? Related: https://github.com/mobile-dev-inc/maestro/issues/1061?

**Comment by artem888** [comment]
> Hey! Great to see this PR, I have some general questions:
> 
> 1. Is this API change required?
> 2. How it works for other platforms (i.e.: Android)?
> 3. Does this solution works for secureTextEntry? Related: [[v1.27.0] `inputText` failing on `secureTextEntry` #1061](https://github.com/mobile-dev-inc/maestro/issues/1061)?

1. yes, it is required, because we need a textfield element to doubleTap on it on iOS (for copy-paste logic)
2. on Android there will be no change, besides when the new API is used we do the tap on the cell automatically
3. I've tested with secure text-fields on iOS, since this change is iOS only and it works fine there

File: maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt [hunk 1]
```diff
@@ -327,21 +327,43 @@ data class AssertConditionCommand(
...
 
+data class InputTextCommandV2(
+    val text: String,
+    val accessibilityText: String? = null,
+    val id: String? = null,
+    val point: String? = null,
+    val pasteTitle: String? = null
+) : Command {
+    override fun description(): String {
+        return "Input text v2 $text"
```
Comment (axelniklasson) [inline maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt:353]: This is being presented to users, so maybe we should remove the `v2` part? We could also update it to add something indicating that it is inputting text to an element directly, thus indicating it is the v2 version.

```suggestion
        return "Input text $text"
```

File: maestro-orchestra/src/main/java/maestro/orchestra/Orchestra.kt [hunk 2]
```diff
@@ -39,6 +40,7 @@ import maestro.orchestra.yaml.YamlCommandReader
 import maestro.toSwipeDirection
 import maestro.utils.MaestroTimer
 import maestro.utils.StringUtils.toRegexSafe
+import java.awt.SystemColor.text
```
Comment (axelniklasson) [inline maestro-orchestra/src/main/java/maestro/orchestra/Orchestra.kt:43]: Is this import required?

**Comment by artem888** [comment]
> Awesome work on this Artem!
> 
> Some follow up questions:
> 
> * Are there changes required on Maestro Cloud?
> * What tests have been made to ensure the changes in the tap implementation does not introduce any regressions?
> 
> Let's also prepare a change request with documentation changes for this command so it is ready for the next release.

- for maestro cloud I need to check, but probably yes
- testing is still WIP, plan is to rerun all maestro-flow-tester tests + I'm doing couple of extra new ones for RN, Flutter and SwiftUI

**Comment by Leland-Takamine** [comment]
Is it possible to query the currently active element to past into instead of changing the inputText API https://github.com/appium/appium/issues/16419

I know when @dmitry-zaitsev tried this he ran into some issues where reactnative apps didn't report the selected element correctly, but curious if you explored this route @artem888 ?

**Comment by artem888** [comment]
> Is it possible to query the currently active element to past into instead of changing the inputText API [appium/appium#16419](https://github.com/appium/appium/issues/16419)
> 
> I know when @dmitry-zaitsev tried this he ran into some issues where reactnative apps didn't report the selected element correctly, but curious if you explored this route @artem888 ?

yes, I did consider that as a first option. there is a parameter `hasKeyboardFocus` in the public API, but I didn't manage to make it work with private APIs + you still need to query the element somehow + it is not even supported by public API queries

**Comment by Leland-Takamine** [comment]
Got it thanks for the context Artem!

**Comment by Leland-Takamine** [comment]
@artem888 @axelniklasson Shouldn't the API allow for any element selector as opposed to just finding the element by id?

**Comment by artem888** [comment]
> @artem888 @axelniklasson Shouldn't the API allow for any element selector as opposed to just finding the element by id?

@axelniklasson @Leland-Takamine that was also a deliberate decision, since the textfield position can be changed after it is tapped on. with default element selector many options wouldn't work in that case. text fields are normally well defined elements that have either accessibility_id or some text content on them, so it is fine to use only those two.

**Comment by Leland-Takamine** [comment]
Ah yeah that makes sense Artem. Ok let's see if disabling autocorrect is a better alternative and make a decision from there

**Comment by dmitry-zaitsev** [comment]
> I know when @dmitry-zaitsev tried this he ran into some issues where reactnative apps didn't report the selected element correctly, but curious if you explored this route @artem888 ?

Just to confirm - yes, that was the case. Querying a currently focused view did not work on React Native.

On selector for `inputText`, I understand that this is not a breaking change but just wanted to share a use-case we have where it is _preferrable_ that `inputText` does not have selector - we want to verify that a view has auto-focus and so we are starting to do `inputText` without doing any tapping prior to that.

**Comment by artem888** [comment]
closing this PR, since another one with better solution is coming.

from this PR we can reuse the logic in the future to implement:
- `pasteTextFrom` command
- `doubleTap` command

**Comment by ArthurSav** [comment]
https://github.com/mobile-dev-inc/maestro/pull/1119

**Pull request closed without merging**